### PR TITLE
feat: thanking contributor (#2632)

### LINKED ISSUE #2632: Thanking contributors
symfony.com has a box at the bottom of their main page titled "They Help Us Make Symfony" that thanks an apparently-random contributor. We should totally do something like this IMO.

Screen Shot 2019-09-30 at 1 45 28 PM

### PR DESCRIPTION
**Pull request by KeziahMoselle**
This PR aims to add a card which thank random Node.js contributors.

## To Do

- [x] Contributor card UI
	- [x] Show loading spinner
- [x] Fetch 1 random contributor
- [x] Optimize / Errors handling / Refactor
	- [x] Handle requests errors
	- [x] Use IntersectionObserver if available
	- [x] Cache the response of the first request in localStorage and update it every month
- [x] Use `s=80` for the avatar image
- [x] Remove the period before the contributions (maybe wrap it in parens)
- [x] Maybe use `textContent` instead of `innerText`?
- [x] Move `.hidden` to a general utility class
- [x] Fix IE >= 10 support
- [x] Fix IE <= 10 styling

## Preview

preview

Fixes #2632 

Live preview: <https://amazing-volhard-25db2f.netlify.com/en/>

### CODE REVIEW

**Comment by KeziahMoselle** [comment]
Update :
preview

**Comment by XhmikosR** [comment]
As I recently learned the hard way, the CSS is currently used by other nodejs sites.

So, this shouldn't land until that's solved.

Also, I'm not very keen of using non standard attributes even if Chrome has the biggest market share.

File: layouts/partials/footer.hbs
Comment (XhmikosR) [inline layouts/partials/footer.hbs]: Also this should be move in main.js and be ES5

**Comment by KeziahMoselle** [comment]
Thank you, i'll change that tomorrow !

File: layouts/partials/footer.hbs
Comment (XhmikosR) [inline layouts/partials/footer.hbs]: Extra `div`

File: layouts/partials/footer.hbs
Comment (KeziahMoselle) [reply layouts/partials/footer.hbs]: Btw does it mean I should use XHR and not fetch at all ?

Comment (XhmikosR) [reply layouts/partials/footer.hbs]: Probably yes. The frontend JS is ES5

File: layouts/partials/footer.hbs
Comment (KeziahMoselle) [reply layouts/partials/footer.hbs]: It's for the display flex (2 childs)

Comment (XhmikosR) [reply layouts/partials/footer.hbs]: Well, divs without classes is bad IMO. Also see my comment about style.css being used by other nodejs sites, which means you can't do such changes.

File: layouts/partials/footer.hbs
Comment (KeziahMoselle) [reply layouts/partials/footer.hbs]: Ok, i'll get rid of that

File: layouts/partials/footer.hbs
Comment (KeziahMoselle) [reply layouts/partials/footer.hbs]: Mmh ok, what can I do ?

Comment (XhmikosR) [reply layouts/partials/footer.hbs]: Not add new elements, or wait until we solve the CSS issue in the other repos.

Comment (KeziahMoselle) [reply layouts/partials/footer.hbs]: I mean, it will be hard to add the contributor card without adding elements, I guess I will wait :)

Comment (XhmikosR) [reply layouts/partials/footer.hbs]: Sure, but this should be ES5 and use elements with proper classes and everything. Empty `div`s is pretty bad.

Comment (KeziahMoselle) [reply layouts/partials/footer.hbs]: You are right

File: layouts/partials/footer.hbs
Comment (XhmikosR) [inline layouts/partials/footer.hbs]: BTW this should be `noreferrer` not `noreff`

File: layouts/css/page-modules/_linuxfoundation.styl
Comment (XhmikosR) [inline layouts/css/page-modules/_linuxfoundation.styl]: You could move these 2 to the HTML

File: layouts/css/page-modules/_linuxfoundation.styl
Comment (XhmikosR) [inline layouts/css/page-modules/_linuxfoundation.styl]: And remove these extra newlines

File: layouts/partials/footer.hbs
Comment (KeziahMoselle) [reply layouts/partials/footer.hbs]: Nice catch, I forgot to remove the second rel

File: layouts/partials/footer.hbs
Comment (XhmikosR) [reply layouts/partials/footer.hbs]: Please rebase from upstream/master and move any JS to static/js/main.js

File: layouts/partials/footer.hbs
Comment (XhmikosR) [inline layouts/partials/footer.hbs]: I'd move width and height after alt.

File: layouts/partials/footer.hbs
Comment (KeziahMoselle) [reply layouts/partials/footer.hbs]: Okay !

**Comment by phillipj** [comment]
Woohoow this is cool, looking forward to seeing this in action 💯

**Comment by XhmikosR** [comment]
FYI that's not a proper rebase

**Comment by KeziahMoselle** [comment]
Oops, I'm sorry it's my first rebase ever, what should i've done ?

**Comment by XhmikosR** [comment]
Assuming you have the upstream repo as upstream, `git rebase -i upstream/master`

**Comment by KeziahMoselle** [comment]
Ok thank's ! I did it without the interactive flag

**Comment by KeziahMoselle** [comment]
Update (after HTML cleanup) :
preview

**Comment by XhmikosR** [comment]
This is still not properly rebased.

**Comment by XhmikosR** [comment]
Also do note `height: 0` is bad. IMO you should occupy the exact dimensions to avoid reflows.

**Comment by XhmikosR** [comment]
Here is the branch with just the current patch https://github.com/nodejs/nodejs.org/compare/master...XhmikosR:thanks-contrib

**Comment by XhmikosR** [comment]
Alright, I pushed a few more tweaks. Live preview <https://pedantic-mclean-3d4d4b.netlify.com/en/>

Still needs polishing, and most importantly see if this will break https://benchmarking.nodejs.org/ or https://coverage.nodejs.org/ which from a quick look it doesn't seem so.

@KeziahMoselle I suggest that you overwrite your branch with mine and keep working on that.

**Comment by KeziahMoselle** [comment]
Thank you for your help !

**Review by ghost — COMMENTED** [review]
Nice feature!!! +100

> Update :
> preview

I tend to use this because I think it will balance the layout with a random image, however if I use something like 
image, which makes me feel making head to the wall because the left side is too crawded. That's only my opition ;)

**Comment by XhmikosR** [comment]
This needs to wait until we have PR previews. And see if we can make only one API request instead of 2, see if we should cache this or not and so on.

Currently this slows down the page a lot that I'm personally not sure if we should move with this.

**Comment by phillipj** [comment]
> Currently this slows down the page a lot that I'm personally not sure if we should move with this.

How do you see this slowing down the page a lot?

To me it looks like the JavaScript introduced here is mostly asynchronous, in other words not blocking rendering. And the Intersection usage supported by [near 90% of users](https://caniuse.com/#search=Intersection) even tries to avoid triggering much of the added code until the `<footer>` is getting close to the viewport. Therefore I've convinced myself it shouldn't have much impact on slowing down the rendering or execution of the other parts of the website.

But you might be considering something that hasn't caught my eye yet 🤷‍♂

**Comment by KeziahMoselle** [comment]
Adding the intersection observer was the idea of @XhmikosR, the home page is slowed down a lot by these 2 requests even with the IO :/

Should I cache the response of the first request to localStorage if supported ? If so how long ?

**Comment by phillipj** [comment]
> the home page is slowed down a lot by these 2 requests even with the IO

This makes me really curious.. Could you describe concretely what you mean by "slowing down" and "a lot"? E.g. is it the actual rendering of the entire webpage we're talking about here? How are you measuring before / after times?

**Comment by XhmikosR** [comment]
This makes 2 XHR requests, and also fetches the avatars.

All this doesn't come for free :) An example https://gtmetrix.com/compare/OUjbjkwL/BXCUZpol

~~Note that the Netlify preview doesn't have the IO patch.~~

**Comment by phillipj** [comment]
Yeay, great to have some insights to base discussions around 👍

Personally think the fact that it takes more time to wait for the browser to finish all its potential HTTP requests (+ followed by some code execution) is less important, and surely not a surprise as you're hinting at.

What is actually important is how quickly visitors gets the website rendered onto their screen. The fact that some parts gets rendered later or even not at all because of network failures etc, may or may not be a problem, it completely depends on *what* gets rendered later. What Google says about [User-centric Performance](https://developers.google.com/web/fundamentals/performance/user-centric-performance-metrics#user-centric_performance_metrics) is spot on IMO:

> Web pages almost always have parts that are more important than others. If the most important parts of a page load quickly, the user may not even notice if the rest of the page doesn't.

In this particular scenario we're talking about nice-to-have content in the website's footer. Surely not *critical* for the visitor to do what they came to the website to do. The rendering of the main content will surely not be postponed significantly because of the changes added here.

Chrome's Performance-part of Devtools shows a few important stages related to User-centric Performance metrics that's really valuable. This is the results I got when measuring the nodejs.org vs these changes on netlify:

#### nodejs.org

Screenshot 2019-10-09 at 13 43 01

#### netlify

Screenshot 2019-10-09 at 13 42 14

Bottom line; I don't think the changes purposed in this PR will negatively affect nodejs.org visitors.

**Comment by XhmikosR** [comment]
I still don't think this is ready.

1. the page jumps when the new div is added; we should have a loading indicator and occupy the space to avoid the page jump
2. I updated the comparison with the latest patch from this branch; IO doesn't affect the main page because of the sticky footer and thus the div being visible there.

**Comment by KeziahMoselle** [comment]
> the page jumps when the new div is added; we should have a loading indicator and occupy the space to avoid the page jump

Yep it's definitely planned !
As @MaledongGit said, I would rather see the card to the right of the paragraphs (this avoid page jump and we can show a loading icon in the middle of the card)

But I don't know if it's possible as you said in your comment here : 
https://github.com/nodejs/nodejs.org/pull/2635#issuecomment-537245772
How can we deal with this ?

**Comment by XhmikosR** [comment]
I will try to come up with a solution for that in the next days. I assume we have only 2 cases where the CSS from nodejs.org is used:

1. https://benchmarking.nodejs.org/
2. https://coverage.nodejs.org/

I will strip the CSS to the minimum and submit patches to those projects. If there are more sites that use the main CSS, now it's a good time to let me to know :)

So, let's put this on hold until this is sorted out.

**Comment by KeziahMoselle** [comment]
Here is a preview with the IntersectionObserver : 
https://condescending-hypatia-7bc62c.netlify.com/en/

And the report : https://gtmetrix.com/compare/OUjbjkwL/BXCUZpol/ksPHx3j1/kefPCt14
1. Without IO
2. Prod
3. Home page with IO
4. About page with IO

**Comment by XhmikosR** [comment]
I have already included it in my preview above, see https://github.com/nodejs/nodejs.org/pull/2635#issuecomment-539974715

**Comment by KeziahMoselle** [comment]
You're right, didn't see that sorry :)

**Comment by XhmikosR** [comment]
Alright, I made 2 PRs

* https://github.com/nodejs/benchmarking/pull/294
* https://github.com/nodejs/build/pull/1947

The first is ready, the second one needs changing the Python script to include styles.css.

That being said, after those land and are deployed, assuming there are no other sites that use the nodejs.org CSS, we should be OK here.

**Comment by KeziahMoselle** [comment]
Awesome ! :tada:

**Comment by KeziahMoselle** [comment]
I will continue to work on this soon, I was very busy lately

**Comment by XhmikosR** [comment]
Note that we still need the coverage PR merged.

**Comment by KeziahMoselle** [comment]
Sure !

**Comment by KeziahMoselle** [comment]
Should I create a new `.scss` file for the card ? If not where can I write the card's styles ?

**Comment by XhmikosR** [comment]
Do note that the [build PR](https://github.com/nodejs/build/pull/1947) hasn't landed yet unfortunately.

**Comment by KeziahMoselle** [comment]
Yes I saw that, I just wanted to know where I can put my styles for now ?

**Comment by XhmikosR** [comment]
A new partial of your choice. We can always tweak all this stuff before merging.

**Comment by KeziahMoselle** [comment]
Okay I will do that ! Thank you 👍

**Comment by KeziahMoselle** [comment]
preview

**Comment by KeziahMoselle** [comment]
Can we replace the avatar by a loading spinner when it fetches ?

**Comment by XhmikosR** [comment]
If it's a pure CSS spinner, why not. Remember to always add width and height to the container to prevent reflows.

**Comment by KeziahMoselle** [comment]
What should we display when there was an error ?
- Replace the `Thank you ...` by an error message
- Remove the card (reflow)
- Other ?

**Comment by ghost** [comment]
@KeziahMoselle : For error happened, "Remove the card (reflow)" is better (through F12 we'll see the result, where's the problem).

BTW: This is welcomed by most of us, so I opened it for a standard feature, @KeziahMoselle will SOON finish his great contribution! :)

**Comment by KeziahMoselle** [comment]
Thank you ! 😄

**Comment by ghost** [comment]
@nodejs/website-redesign: Any other suggestions or ideas, it's fine on me.

**Comment by Trott** [comment]
@nodejs/website PTAL

**Comment by bnb** [comment]
I love this as a concept. As we’re developing the redesign we should also be sure to include similar functionality there - cc @nodejs/website-redesign

**Review by osk2 — APPROVED** [review]
Awesome, can't wait to see it landed!

**Review by ryzokuken — APPROVED** [review]
❤️Thanks for doing this!

RSLGTM.

**Comment by Trott** [comment]
Anyone up for doing the rebase that will be required to land this?

**Comment by KeziahMoselle** [comment]
As @XhmikosR said, I think we still need to wait until this PR : https://github.com/nodejs/build/pull/1947 is merged

**Comment by zlatanvasovic** [comment]
@Trott doesn't GitHub allow automatic merging for this one? Anyway if you want to squash all the commits, I can do it.

**Comment by Trott** [comment]
> @Trott doesn't GitHub allow automatic merging for this one? 

Usually, but not in this case:

> This branch cannot be rebased due to conflicts
> Rebasing the commits of this branch on top of the base branch cannot be performed automatically due to conflicts encountered while reapplying the individual commits from the head branch.

**Comment by zlatanvasovic** [comment]
@Trott rebase doesn't seem to be required though. Squash does it the nice way.
Screenshot from 2019-11-26 01-01-24

**Comment by Trott** [comment]
> @Trott rebase doesn't seem to be required though. Squash does it the nice way.
> Screenshot from 2019-11-26 01-01-24

Squash will not work here any more than rebase because of merge conflicts that need to be resolved manually. The squash/merge/rebase button is disabled on this PR for that reason. I haven't looked to see if the manual rebase required to get rid of the merge conflict is a complicated one or not. It might be because there are 24 commits so there's a lot of possibilities for a cascade of conflicts. Or it might be a simple single conflict that needs to be fixed. I just wanted to know if there was someone who was already the point person for getting rid of the merge conflict or not.

**Comment by Trott** [comment]
> > @Trott rebase doesn't seem to be required though. Squash does it the nice way.
> > Screenshot from 2019-11-26 01-01-24
> 
> Squash will not work here any more than rebase because of merge conflicts that need to be resolved manually. The squash/merge/rebase button is disabled on this PR for that reason. I haven't looked to see if the manual rebase required to get rid of the merge conflict is a complicated one or not. It might be because there are 24 commits so there's a lot of possibilities for a cascade of conflicts. Or it might be a simple single conflict that needs to be fixed. I just wanted to know if there was someone who was already the point person for getting rid of the merge conflict or not.

Nope, wait, I'm wrong, squash is enabled. I blame the UI for my confusion. OK, well, cool, that'll work (as long as squashing is OK here--looks like there are multiple authors of the commits).

**Review by aymen94 — APPROVED** [review]

**Comment by XhmikosR** [comment]
I rebased this properly (wasn't easy), but I can't push it for some reason...

>git push --force-with-lease "kf" feat-thanking-contributors:feat-thanking-contributors
remote: Permission to KeziahForks/nodejs.org.git denied to XhmikosR.
fatal: unable to access 'https://github.com/KeziahForks/nodejs.org.git/': The requested URL returned error: 403

@KeziahMoselle please check if you have unchecked the PR option Allow Edit from maintainers.

Notes:

1. Not sure about squashing this yet
2. We are blocked due to https://github.com/nodejs/build/pull/1947
3. The period at the end seems weird?

**Comment by KeziahMoselle** [comment]
It's checked : 
img

**Comment by XhmikosR** [comment]
Not sure why it doesn't allow me to push. Have you protected the branch maybe?

**Comment by KeziahMoselle** [comment]
I didn't do anything special to my fork, I added you as a collaborator, try if it works now ?

**Comment by XhmikosR** [comment]
Yeah, now it works, but it should have worked without me having push rights to your fork.

Anyway, the branch is rebased now, but see my points above https://github.com/nodejs/nodejs.org/pull/2635#issuecomment-558609100

**Comment by KeziahMoselle** [comment]
It's maybe because I forked this repo in an organization and not under my personal account ?

**Comment by XhmikosR** [comment]
>It's maybe because I forked this repo in an organization and not under my personal account ?

Yeah, probably that's the cause.

Live preview: <https://amazing-volhard-25db2f.netlify.com/en/>

Also, when the username is long, the text wraps below the image. Not sure if that's intended.

Now, performance-wise, this is still a lot slower due to the 2 sequential API calls https://gtmetrix.com/compare/luyaCLIo/0XL8kYFr#waterfall

We are also loading one extra font file as it seems, which my bet is that it's due to `strong` thus font-weight 700.

That being said, I'll leave the final decision to the others. :) It's a nice patch, I just want to point out the pros and cons and see what we can do to improve things.

PS. note that the size comparison isn't fair since netlify doesn't use brotli and they don't seem to compress favicon.ico like Cloudflare does.

**Comment by XhmikosR** [comment]
Now, performance-wise, this is still a lot slower due to the 2 sequential API calls https://gtmetrix.com/compare/luyaCLIo/0XL8kYFr#waterfall

We are also loading one extra font-weight as it seems, which my bet is that it's due to `strong` thus font-weight 700.

That being said, I'll leave the final decision to the others. :) It's a nice patch, I just want to point out the pros and cons and see what we can do to improve things.

PS. note that the size comparison isn't fair since netlify doesn't use brotli and they don't seem to compress favicon.ico like Cloudflare does.

**Comment by phillipj** [comment]
Great job pushing this forward 💯 

I haven't spent time understanding what sounds like style blockers in the build-repo, so not opinionated about that part.

Although I'm still convinced the additional HTTP requests involved here is no issue at all, so that's not even near being a con in my eyes. Understand why it might seem like an issue in a waterfall graph of HTTP requests, but that means not considering the bigger picture of how browsers works and to what degree it will degrade the end-user experience in practise as I was trying to explain in October (https://github.com/nodejs/nodejs.org/pull/2635#issuecomment-539970154).

LGTM as soon as external blockers has been resolved.

**Comment by XhmikosR** [comment]
>I haven't spent time understanding what sounds like style blockers in the build-repo, so not opinionated about that part.

I learned this the hard way by breaking things a few months ago, that the main CSS file from nodejs.org is unfortunately used by other nodejs sites. I was told that two were doing this, https://coverage.nodejs.org/ and https://benchmarking.nodejs.org/. For benchmarking, my PR to stop using the CSS from nodejs.org was merged and deployed. For coverage it's still pending.

And this kind of changes can break the layout in the other sites hence why it's blocking.

The slowness is from the sequential API calls we make here. The extra font is not a blocker and we could even work around it if needed.

**Comment by XhmikosR** [comment]
BTW personally I have expressed my opinion that no site should share assets, but it seems no one agreed on moving the rest of the shared assets like favicons and the logo. Which means things will break again if any of these files change :P

**Comment by phillipj** [comment]
> The slowness is from the sequential API calls we make here.

Do you see those requests degrading the user experience from what we have today? Or are you simply stating the fact that more HTTP requests are performed from a technical point of view, and the time it takes for the browser to complete all of those will take longer?

**Comment by XhmikosR** [comment]
Well, this depends. For people with slow internet it will be noticeable since we are talking about a 5x slowness for the page to finish loading. But being that we do show a loader and the layout doesn't change while loading AFAICT, I think the user experience isn't bad.

**Comment by phillipj** [comment]
> BTW personally I have expressed my opinion that no site should share assets, but it seems no one agreed on moving the rest of the shared assets like favicons and the logo.

Sounds sensible to me 👍 Even more so when we've seen the current approach explode.

Did you suggest moving in that direction as part of a PR or similar?

**Comment by XhmikosR** [comment]
> Did you suggest moving in that direction as part of a PR or similar?

https://github.com/nodejs/benchmarking/pull/294#issuecomment-541303341

and

https://github.com/nodejs/build/pull/1947

**Comment by phillipj** [comment]
> For people with slow internet it will be noticeable since we are talking about a 5x slowness for the page to finish loading.

This is where seeing the bigger picture of how the browser works is absolutely crucial.

Without rambling about all the different ready- & loaded events browsers has, it's safe to say that those end-users will have the majority of our webpage' content loaded & rendered to them a long time before the requests added by this PR finishes.

Do you agree on that point or see it differently?

**Comment by XhmikosR** [comment]
Yeah, the main part will always be loaded first. And if the page has long content and the browser supports IntersectionObserver, it will be lazy loaded until the thanking contributor part is in the viewport.

https://www.webpagetest.org/video/compare.php?tests=191126_FD_57aabadf06f68c3e4018e5d331681f8a,191126_A4_1c8120ec9bb5209444e0d7034735a986

The above example is actually the case I described above; the footer is not visible so it doesn't really affect the page. The slowness is just from using netlify AFAICT and maybe the extra font, which the latter we can work around it.

**Review by 3imed-jaberi — APPROVED** [review]

File: static/js/main.js
Comment (XhmikosR) [inline static/js/main.js]: We should append `&s=80` since we know we want this to be 40x40, thus it will still look good on retina

File: static/js/main.js
Comment (XhmikosR) [inline static/js/main.js]: Should add `nofollow noopener noreferrer` here too

File: static/js/main.js
Comment (XhmikosR) [inline static/js/main.js]: This can be probably merged with the line before

**Comment by XhmikosR** [comment]
I added 3 more TODO, so let's get these sorted :)

- [ ] Use `s=80` for the avatar image
- [ ] Remove the period before the contributions (maybe wrap it in parens)
- [ ] Maybe use `textContent` instead of `innerText`?

**Comment by KeziahMoselle** [comment]
You're right, it wasn't intended !
Can I use flexbox to solve this issue ?

**Comment by ghost** [comment]
@KeziahMoselle: Conflicted files....

**Comment by XhmikosR** [comment]
@MaledongGit please do not attempt to merge this. There are pending stuff that need to be addressed. I will rebase this myself, but unless the remaining issues are solved, this should be marked on hold.

**Comment by phillipj** [comment]
Would it be an idea to use GitHub's draft status for this PR to make that
point of *not* merging as explicit as possible?

**Comment by XhmikosR** [comment]
You can't make a PR draft after it's opened as a normal one.

**Comment by phillipj** [comment]
Ahh thanks, wasn't aware of that limitation.

**Comment by KeziahMoselle** [comment]
How can we distinguish the contributions ?

**Comment by KeziahMoselle** [comment]
> [83bda33](https://github.com/nodejs/nodejs.org/commit/83bda33576aa41c031db33a7170f4a8f3120c543)

What did you mean ? @gaiguy18

**Comment by XhmikosR** [comment]
This PR needs to either be locked or we need a new one that is made as a draft. I can't keep rebasing to fix wrong merges.

**Comment by XhmikosR** [comment]
Alright, I fixed a few IE issues, but everything is still broken on IE >= 10.

ie10

IE 11 does not have the weird layout issue which my bet is due to a flexbox bug, but still shows undefined.

**Comment by WaleedAshraf** [comment]
@KeziahMoselle where does it fetch contributors from?

and amazing work 👍 🎉

**Comment by KeziahMoselle** [comment]
> @KeziahMoselle where does it fetch contributors from?

@WaleedAshraf It fetches a random contributor from the `nodejs/node` repository via the GitHub API v3, which means two thing :
1. I can't fetch the 2,625 current contributors (only ~500 on this repository)
2. I can't fetch contributors from all nodejs/ repositories

If anyone know why the GitHub API returns only ~500 contributors and not the real number, I would like to know 😄

([This function](https://github.com/nodejs/nodejs.org/pull/2635/files#diff-93471d4bf84d0569fadcebe5fba72773R113) fetch the max number of contributors and call `getContributor` with a random number)

**Comment by XhmikosR** [comment]
Alright, I rebased this and also I fixed IE 10 JS. There are still a few issues with the avatar and text overflowing the card on IE but I don't have the time to find a way to fix it. Maybe later or someone else could help :)

IE10:

ie10

IE11:

ie11

Live preview: <https://amazing-volhard-25db2f.netlify.com/en/>

**Comment by KeziahMoselle** [comment]
Awesome ! Thank you so much for that.
I don't have much knowledge about the support of IE10 so I don't think I can help with that. I'm sorry !

**Comment by XhmikosR** [comment]
I'll try to have a look in the next days but I can't tell if I'll be able to solve this. That being said, at least it's working now. Some final tweaks might be needed and then we can land this I think :)

**Comment by KeziahMoselle** [comment]
Yep ! I think we still need to put forward the contributions ?

**Comment by XhmikosR** [comment]
I pushed a few more tweaks. Now IE11 seems to be fine; only IE 10 and below have issues, see https://github.com/nodejs/nodejs.org/pull/2635#issuecomment-569019602

**Review by keywordnew — APPROVED** [review]
Thank you, LGTM 👏🏽

**Comment by keywordnew** [comment]
This is such a good feature!

It would be wonderful to have this also ported to the Website Redesign initiative: https://github.com/nodejs/nodejs.dev

**Comment by XhmikosR** [comment]
Rebased. I could use some help for the remaining IE issues, otherwise let's land this as is, without squashing the patches.

**Comment by XhmikosR** [comment]
@MartijnCuppens I don't suppose you have some time to help us out with https://github.com/nodejs/nodejs.org/pull/2635#issuecomment-569019602 please? :)

**Comment by MartijnCuppens** [comment]
@XhmikosR, I've opened https://github.com/KeziahForks/nodejs.org/pull/1

File: layouts/css/page-modules/_contributor-card.scss
Comment (XhmikosR) [inline layouts/css/page-modules/_contributor-card.scss]: @MartijnCuppens maybe this was supposed to be `padding-left`?

Comment (MartijnCuppens) [reply layouts/css/page-modules/_contributor-card.scss]: Yeah, indeed

**Comment by XhmikosR** [comment]
Thanks @MartijnCuppens!

Preview: <https://amazing-volhard-25db2f.netlify.com/en/>

I think this is ready to be merged as a first patch. Just remember to not squash the patches.

PS. There might be room for improvements later.

**Comment by XhmikosR** [comment]
Crap, forgot that we are still blocked by https://github.com/nodejs/build/pull/1947

**Comment by Trott** [comment]
> Just remember to not squash the patches.

Why not? What's the motivation? Is it merely to prefer authorship? If so, I'd prefer we squash and add [co-author metadata](https://help.github.com/en/github/committing-changes-to-your-project/creating-a-commit-with-multiple-authors). We don't need 20+ commits here. Doing so might break `git bisect` for some cases. If the commits were logically separated into independent changes, not squashing would make sense. But they're not.

**Comment by XhmikosR** [comment]
>Is it merely to prefer authorship

This part is not a trivial one. It's quite important. So, yeah, I do want to keep them separate. I couldn't squash more patches because there are conflicts, and this PR is the result of a lot of work.

Either way, we really need that build PR to be done so that we untie the main repo from the other ones.

**Comment by Trott** [comment]
> > Is it merely to prefer authorship
> 
> This part is not a trivial one.

Didn't mean to suggest it was trivial, just that there were other mechanisms for it if that was the only concern. Sorry if "merely" came off the wrong way there!

Maybe we could do a compromise by rebasing and squashing adjacent commits by the same author into each other? So it's only 7 commits instead of 21?

**Comment by XhmikosR** [comment]
>Maybe we could do a compromise by rebasing and squashing adjacent commits by the same author into each other? So it's only 7 commits instead of 21?

That I can try for sure.

But see if we can get someone to help us move with the build PR.

**Comment by KeziahMoselle** [comment]
Thank you for your help @MartijnCuppens and @XhmikosR !
I'm sorry if I commited errors that you needed to fix 🙏

**Review by ghost — APPROVED** [review]
Can this be merged?

**Comment by XhmikosR** [comment]
@MaledongGit no. Please don't bother with this PR. I will merge it when we are ready.

**Comment by KeziahMoselle** [comment]
Thank you so much for merging this PR !
A big thanks to @XhmikosR who helped a lot on this 🙏 
Thank's to all the reviewers and the Node team, keep it up !
This merging made my day !

**Comment by XhmikosR** [comment]
Thanks for sticking with the patch till the end :)

**Comment by tniessen** [comment]
> If so, I'd prefer we squash and add co-author metadata.

:+1:

---

This breaks if going to a 404 URL, because the language picker function throws, and this code never runs, leaving the user with a spinner and placeholder text.

**Comment by XhmikosR** [comment]
@tniessen can you share the stacktrace please? Because when developing locally I don't get the 404 page.

**Comment by XhmikosR** [comment]
BTW ideally you should make a new issue.

**Comment by XhmikosR** [comment]
NVM, I saw the error on the live site. It's unrelated to this PR so please make a new issue. It has to do with the language toggler.

**Comment by Trott** [comment]
Issue for toggler code failure: https://github.com/nodejs/nodejs.org/issues/2984

**Pull request merged**
